src/qibolab/instruments/zhinst/executor.py

```diff
@@ -0,0 +1,781 @@
+                    )  # Probability inversion patch
+                serial = ropulse.pulse.serial
+                qubit = ropulse.pulse.qubit
+                results[serial] = results[qubit] = options.results_type(data)
+
+        return results
+
+    def sequence_zh(
+        self, sequence: PulseSequence, qubits: dict[str, Qubit]
+    ) -> dict[str, list[ZhPulse]]:
+        """Convert Qibo sequence to a sequence where all pulses are replaced
+        with ZhPulse instances.
+
+        The resulting object is a dictionary mapping from channel name
+        to corresponding sequence of ZhPulse instances
+        """
+        # Define and assign the sequence
+        zhsequence = defaultdict(list)
+
+        # Fill the sequences with pulses according to their lines in temporal order
+        for pulse in sequence:
+            if pulse.type == PulseType.READOUT:
+                ch = measure_channel_name(qubits[pulse.qubit])
+            else:
+                ch = pulse.channel
+            zhsequence[ch].append(ZhPulse(pulse))
+
+        if self.processed_sweeps:
+            for ch, zhpulses in zhsequence.items():
+                for zhpulse in zhpulses:
+                    for param, sweep in self.processed_sweeps.sweeps_for_pulse(
+                        zhpulse.pulse
+                    ):
+                        zhpulse.add_sweeper(param, sweep)
+
+        return zhsequence
+
+    def create_exp(self, qubits, options):
+        """Zurich experiment initialization using their Experiment class."""
+
+        # Setting experiment signal lines
+        signals = [lo.ExperimentSignal(name) for name in self.signal_map.keys()]
+
+        exp = lo.Experiment(
+            uid="Sequence",
+            signals=signals,
+        )
+
+        if self.acquisition_type:
+            acquisition_type = self.acquisition_type
+            self.acquisition_type = None
+        else:
+            acquisition_type = ACQUISITION_TYPE[options.acquisition_type]
+        averaging_mode = AVERAGING_MODE[options.averaging_mode]
+        exp_options = replace(
+            options, acquisition_type=acquisition_type, averaging_mode=averaging_mode
+        )
+
+        # Near Time recursion loop or directly to Real Time recursion loop
+        if self.nt_sweeps:
+            self.sweep_recursion_nt(qubits, exp_options, exp)
+        else:
+            self.define_exp(qubits, exp_options, exp)
+
+    def define_exp(self, qubits, exp_options, exp):
+        """Real time definition."""
+        with exp.acquire_loop_rt(
+            uid="shots",
+            count=exp_options.nshots,
+            acquisition_type=exp_options.acquisition_type,
+            averaging_mode=exp_options.averaging_mode,
+        ):
+            # Recursion loop for sweepers or just play a sequence
+            if len(self.rt_sweeps) > 0:
+                self.sweep_recursion(qubits, exp, exp_options)
+            else:
+                self.select_exp(exp, qubits, exp_options)
```

Comment:
Duplicated in:
https://github.com/qiboteam/qibolab/blob/c982befc9865f35d86f046c9cabb31c8404c171c/src/qibolab/instruments/zhinst/executor.py#L739-L742

Same of above.